scripts/phpmqtt_input.php

```diff
@@ -121,12 +121,24 @@
         if (!$connected && (time()-$last_retry)>5.0) {
             $last_retry = time();
             try {
+                // SUBSCRIBE
                 $mqtt_client->setCredentials($mqtt_server['user'],$mqtt_server['password']);
                 $mqtt_client->connect($mqtt_server['host'], $mqtt_server['port'], 5);
                 $topic = $mqtt_server['basetopic']."/#";
                 //echo "Subscribing to: ".$topic."\n";
                 $log->info("Subscribing to: ".$topic);
                 $mqtt_client->subscribe($topic,2);
+
+                // PUBLISH
```

Comment:
@TrystanLea - doesn't seem to loop continually? only grabs the redis data on "first run"... any ideas?

Reply:
Is it looping the whole queue in the first run or just the first item in the queue?

I suspect the former, I think the publish will only get called once unless there are some connection issues.

I'm not convinced that is the right place for the publish. Surely it needs to be in the outer loop does it not? Where it is now it will only get called when there is no connection and it's been at least 5s since the last attempted reconnection.

Also I think you need to loop faster and just pop the first item in the queue each loop otherwise if there is a backlog of data to be published it will block new inputs arriving. eg after a reconnection and a flood of processing ends up with hundreds or publish to mqtt's, if you loop through the whole list it will take a while. This is especially important as the incoming data is not currently timestamped until emoncms processes it so any delays will alter the datapoint timestamps, or if there is a disconnection (possibly due to too many publishes) the incoming data queued at the broker will be lost as emoncms will reconnect with a new client id, ie clean slate.

Incoming data should take priority over outgoing at this stage.